File: problems/2026/04/2026-04-16-leetcode-206-反转链表/SolutionLC206.py
from __future__ import annotations

from typing import List, Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None):
        self.val = val
        self.next = next


class Solution:
    """LeetCode 206：反转链表（迭代三指针）"""

    def reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev: Optional[ListNode] = None
        curr = head
        while curr is not None:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        return prev


def _from_vals(vals: List[int]) -> Optional[ListNode]:
    if not vals:
        return None
    head = ListNode(vals[0])
    p = head
    for x in vals[1:]:
        p.next = ListNode(x)
        p = p.next
    return head


def _to_vals(head: Optional[ListNode]) -> List[int]:
    out: List[int] = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


if __name__ == "__main__":
    sol = Solution()
    assert _to_vals(sol.reverse_list(_from_vals([1, 2, 3, 4, 5]))) == [5, 4, 3, 2, 1]
    assert _to_vals(sol.reverse_list(_from_vals([1, 2]))) == [2, 1]
    assert _to_vals(sol.reverse_list(None)) == []
    print("示例均通过")

File: problems/2026/04/2026-04-16-leetcode-206-反转链表/SolutionLC206.java
/**
 * LeetCode 206：反转链表（迭代三指针，O(1) 额外空间）
 */
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

public class SolutionLC206 {

    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode nxt = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nxt;
        }
        return prev;
    }

    private static ListNode build(int... vals) {
        if (vals.length == 0) {
            return null;
        }
        ListNode head = new ListNode(vals[0]);
        ListNode p = head;
        for (int i = 1; i < vals.length; i++) {
            p.next = new ListNode(vals[i]);
            p = p.next;
        }
        return head;
    }

    private static String toStr(ListNode head) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        ListNode p = head;
        while (p != null) {
            if (sb.length() > 1) {
                sb.append(",");
            }
            sb.append(p.val);
            p = p.next;
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        SolutionLC206 solution = new SolutionLC206();
        ListNode a = build(1, 2, 3, 4, 5);
        System.out.println(toStr(solution.reverseList(a)));

        ListNode b = build(1, 2);
        System.out.println(toStr(solution.reverseList(b)));

        System.out.println(toStr(solution.reverseList(null)));
    }
}
